examples/codes/external_code.ex.py

```diff
@@ -0,0 +1,428 @@
+        )
+        self._teardown = self.teardown_cls(self.params, build_config["outfile"])
+
+    def execute(self, run_mode: Union[str, RunMode]) -> ParameterFrame:
+        """Execute the solver"""
+        if isinstance(run_mode, str):
+            run_mode = self.run_mode_cls.from_string(run_mode)
+        result = None
+        if setup := self._get_execution_method(self._setup, run_mode):
+            result = setup()
+        if run := self._get_execution_method(self._run, run_mode):
+            run(result)
+        if teardown := self._get_execution_method(self._teardown, run_mode):
+            result = teardown()
+        return result
+
+
+# %% [markdown]
+# ### Using the solver
+#
+# To run the solver you just need to provide the parameters and the configuration
+# to initialse the object.
+# Be aware `problem_settings` should be used sparingly for options that won't
+# change within the rerunning of the solver,
+# it has the same effect as modifying the default.
+# Also note the units of the parameter values returned back have been updated.
+#
+# The files written and read by the external code are stored in the generated_data
+# folder in the root of the bluemira repository.
+#
+# Some warnings will be shown because some of the situations here are usually
+# undesireable.
+# In this first block we will see 3 warnings.
+# The first 2 are the same for the `run` and `read` modes:
+#   * "No value for param1"
+#       - param 1 has its send mapping set to `False` so no value is sent to
+#         the code and therefore it is not read back in. As `read` mode is executed
+#         with the output of the `run` mode the same error is repeated.
+#
+# The 3rd is for the `mock` mode:
+#   * "No value or param2"
+#       - The mock output doesn't have a param2 output
+#
+# Notice that `param2` does not take the value given in problems settings as we
+# overwrite it because `solver.params.mappings['param2'].send == True`,
+# set in the `ECParameterFrame` default mappings.
+
+# %%
+io_path = get_bluemira_path("", "generated_data")
+
+params = {
+    "header": {"value": False, "unit": "", "source": "here"},
+    "param1": {"value": 5, "unit": "W"},
+}
+
+build_config = {
+    "problem_settings": {"param2": 10},
+    "infile": Path(io_path, "infile.txt"),
+    "outfile": Path(io_path, "outfile.txt"),
+}
+solver = Solver(params, build_config)
+
+# Running in all the different modes
+for mode in ["run", "read", "mock"]:
+    print(mode)
+    out_params = solver.execute(mode)
+    print(out_params)
+
+# %% [markdown]
+# 1 warning this time, we still havent sent a value for `param1` in `run` mode.
+# Notice how the default for `param2` from our `problem_settings` is used when we
+# turn off the send mapping
+
+# %%
+solver.modify_mappings({"param2": {"send": False}})
+print(solver.execute("run"))
+
+# %% [markdown]
+# Again the same warning. This time we have modified the value of `param2` and turned
+# the send mapping back on
```

Comment:
```suggestion
# the send mapping back on.
```

Reply:
I'll stop suggesting these now, might be worth going through and ensuring our markdown blocks end with a full-stop where they should